File: src/main/java/com/hedvig/backoffice/domain/QuestionGroup.java
package com.hedvig.backoffice.domain;

import java.time.Instant;
import java.util.HashSet;
import java.util.Set;
import javax.persistence.CascadeType;
import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.GeneratedValue;
import javax.persistence.Id;
import javax.persistence.Lob;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;
import javax.persistence.OrderBy;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.Type;

@Entity
@Getter
@Setter
@NoArgsConstructor
public class QuestionGroup {

  @Id @GeneratedValue public Long id;

  public Instant date;

  public Instant answerDate;

  @Lob
  @Type(type = "org.hibernate.type.TextType")
  public String answer;

  @ManyToOne public Personnel personnel;

  @ManyToOne public Subscription subscription;

  @OneToMany(fetch = FetchType.EAGER, cascade = CascadeType.ALL, mappedBy = "group")
  @OrderBy("date")
  public Set<Question> questions = new HashSet<>();

  public QuestionGroup(Subscription subscription) {
    this.subscription = subscription;
  }

  public void addQuestion(Question question) {
    question.setGroup(this);
    questions.add(question);
  }

  public void correctDate(Instant date) {
    if (this.date != null && this.date.isAfter(date)) {
      return;
    }

    this.date = date;
  }
}

File: src/main/java/com/hedvig/backoffice/domain/QuestionGroup.kt
package com.hedvig.backoffice.domain

import org.hibernate.annotations.Type
import java.time.Instant
import java.util.*
import javax.persistence.*

@Entity
class QuestionGroup(
  @ManyToOne
  var subscription: Subscription
) {
  @Id
  @GeneratedValue
  var id: Long = 0
  var date: Instant? = null
  var answerDate: Instant? = null
  @Lob
  @Type(type = "org.hibernate.type.TextType")
  var answer: String? = null
  @ManyToOne
  var personnel: Personnel? = null
  @OneToMany(fetch = FetchType.EAGER, cascade = [CascadeType.ALL], mappedBy = "group")
  @OrderBy("date")
  var questions: MutableSet<Question> = HashSet()

  fun addQuestion(question: Question) {
    question.group = this
    questions.add(question)
  }

  fun correctDate(date: Instant) {
    if (this.date != null && date.isAfter(date)) {
      return
    }
    this.date = date
  }
}
